arch/arm/arm64/include/uk/asm/paging.h

```diff
@@ -47,19 +65,213 @@ struct ukarch_pagetable {
 };
 #endif /* !__ASSEMBLY__ */
 
+/* ---------------- FIXME ------------------------ */
+/*These are also defined in plat/common/include/arm/arm64/cpu_defs.h */
+#define PTE_Lx_VALID_BIT	1UL
+#define PTE_Lx_TYPE_BIT		(1UL << 1)
+#define PTE_Lx_TYPE_BLOCK	0
+
+#define PTE_L2_BLOCK_PADDR_MASK	_AC(0x0000ffffc0000000, UL) /* [47-30] */
+#define PTE_L1_BLOCK_PADDR_MASK	_AC(0x0000ffffffe00000, UL) /* [47-21] */
+#define PTE_L0_PAGE_PADDR_MASK	_AC(0x0000fffffffff000, UL) /* [47-12] */
+#define PTE_Lx_TABLE_PADDR_MASK	_AC(0x0000fffffffff000, UL) /* [47-12] */
+
+#define DEVICE_nGnRnE	0
+#define DEVICE_nGnRE	1
+#define DEVICE_GRE	2
+#define NORMAL_NC	3
+#define NORMAL_WT	4
+#define NORMAL_WB	5
+
+#define TTBR0_EL1_BADDR_MASK _AC(0x0000fffffffffffe, UL)
+/* ---------------- FIXME ----------------------- */
```

Comment:
Okay, let me try to explain the problem: Currently `cpu_defs.h` contains two types of information:
1. Architectural definitions, ie information provided in the Arm ARM.
2. Definitions defined by the platform. For example the indices we choose to use in MAIR_EL1 for different memory types, or the default PTE attributes we choose to use when we populate the pagetables. I will refer to this below as "policy".

This is problematic because:
1. Register definitions cannot be used by headers under `arch/` (hence the duplication you see above)
2. The semantics of `cpu_defs.h` are not clear. It's a mixed bag.
3. Policy is still required in `arch`, see the definitions of device-type in lines 79-84.

To solve 1 and 2, I would rather have all architectural definitions in a single header under `arch/arm/arm64/`, with a rule that says "If it's not defined in the Arm ARM, it doesn't go here". I think that's much cleaner that spreading architectural definitions across different headers, and it's also a common practice in general.

For 3, I don't know what's the right way to go 😅

Reply:
I agree with the solution to 1 and 2 to move all strict architecture-specific definitions like registers etc. to `arch/ARCH/` and actually never understood why this has not been done right from the start 🤷🏻‍♂️ I think it is ok to have them grouped semantically though, like having PTE bit definitions still be in `paging.h`. As a compromise to a single header, this would also allow them to be mixed with "policy" definitions and still feel right. But at least all arch definitions would be in one directory and not some only available in `plat`, some globally available and (even worse) other ones duplicated.